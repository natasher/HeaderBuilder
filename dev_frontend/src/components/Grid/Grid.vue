<template>
  <div
    class        = "grid"
    :data-status = "getGridStatus">

    <wrapper-clearfix class = "window primary">

      <grid-header />

      <wrapper-clearfix class = "main">

        <action-bar />
        <first-row  />
        <second-row />

      </wrapper-clearfix>

      <div class = "grid-auto" v-if = "getGridStatus === 'auto'">
        You're in Auto Mode
      </div>

    </wrapper-clearfix>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'
import WrapperClearfix from '../BaseGeneric/WrapperClearfix.vue'
import GridHeader      from './Segments/GridHeader.vue'
import ActionBar       from './Rows/ActionBar.vue'
import FirstRow        from './Rows/FirstRow.vue'
import SecondRow       from './Rows/SecondRow.vue'

export default {

  name: 'Grid',

  components: {
    WrapperClearfix,
    GridHeader,
    ActionBar,
    FirstRow,
    SecondRow,
  },

  computed: {
    ...mapGetters( 'ui', [ 'getGridStatus' ]),
  },

}
</script>
