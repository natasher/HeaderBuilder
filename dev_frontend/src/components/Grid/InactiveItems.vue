<template>
<section id = "mfn-hb-inactive-items">

  <div class = "header">
    <span class = "title">
      Inactive items
    </span>
  </div>

  <div class = "hb-items">
    <draggable
      class    = "hb-col"
      v-model  = "inactiveItems"
      key      = "inactiveItems"
      @change  = "onChange"
      :options = "{
        group: {
          name: 'items',
        },
        scroll: true,
        disabled: ( getGridStatus === 'off' || getGridStatus === 'auto')
                  ? true
                  : false,
      }" >

      <base-item
        v-for   = "( item, index ) in inactiveItems"
        row     = "inactiveItems"
        :item   = "item"
        :cellId = "index"
        :key    = "item.uuid" />

    </draggable>
  </div>

</section>
</template>

<script>
import { mapGetters, mapActions } from 'vuex'
import FontIcon  from '../BaseGeneric/FontIcon.vue'
import BaseItem  from '../BaseItem.vue'
import draggable from 'vuedraggable'

export default {

  name: 'InactiveItems',

  components: {
    FontIcon,
    BaseItem,
    draggable,
  },

  computed: {
    ...mapGetters( 'ui', [
      'getGridStatus',
    ]),

    ...mapGetters( 'inactiveItems', [
      'getInactiveItems',
    ]),

    inactiveItems: {
      get() {
        return this.getInactiveItems;
      },
      set() {}
    }
  },

  methods: {
    ...mapActions( 'inactiveItems', [
      'removeItem',
    ]),

    onChange( event ) {

      if ( event.removed ) {
        const cellId = event.removed.oldIndex

        this.removeItem({
          cellId : cellId,
        })
      }

    },
  },

}
</script>
