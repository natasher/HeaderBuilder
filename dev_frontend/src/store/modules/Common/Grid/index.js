export default {

  namespaced: true,

  actions: {

    /**
     * Clone `grid` from given in payload module to currently rendered Module.
     * Begin with clearing all cells to avoid item duplicates.
     * If current view is 'mobile', grid should contain only 'logo' and 'menu',
     * rest of the items should be placed in `inactiveItems`.
     * 'desktop' and 'tablet' should be mapped 1-by-1 as { from } module.
     *
     * @param {function} dispatch Asynchronous invoke action.
     * @param {Object} rootGetters Access getters from global namespace.
     * @param {Object} rootState Access to global state.
     * @param {Object} payload Object that represents data sends to store
     *    - from  {string}:[ 'DesktopModule', 'TabletModule', 'MobileModule' ]
     */
    clone({ dispatch, rootGetters, rootState }, payload ) {
      return new Promise(( resolve ) => {

        [ 'actionBar', 'firstRow', 'secondRow' ].map(( curr_row ) => {
          [ 'left', 'center', 'right' ].map(( curr_cell ) => {

            dispatch( 'cell/clear', {
              row    : curr_row,
              cellPos: curr_cell,
            }, { root: true })
            .then(() => {
              const currView = rootState.general.currentView

              if ( rootGetters[ 'ui/getGridStatus' ] === 'auto' && currView === 'mobile' ) {

                rootState[ payload.from ][ curr_row ][ 'items' ][ curr_cell ]
                  .filter(( item ) => {
                    if ( item.name === 'logo' || item.name === 'menu' ) {

                      return item

                    } else {

                      dispatch( 'inactiveItems/putItem', {
                        item: {
                          name : item.name,
                          icon : item.icon,
                          form : item.form,
                          style: item.style || {},
                          uuid : this._vm.$genid(),
                          // originalCoords: item.originalCoords,
                        }
                      }, { root: true })

                    }
                  })
                  .map(( item ) => {

                    dispatch( 'cell/pushItem', {
                      row    : curr_row,
                      cellPos: curr_cell,
                      item: {
                        name          : item.name,
                        icon          : item.icon,
                        form          : item.form,
                        style         : item.style || {},
                        uuid          : this._vm.$genid(),
                        originalCoords: item.originalCoords,
                      }
                    }, { root: true })

                  })

              } else {

                rootState[ payload.from ][ curr_row ][ 'items' ][ curr_cell ]
                  .map(( item ) => {

                    dispatch( 'cell/pushItem', {
                      row    : curr_row,
                      cellPos: curr_cell,
                      item   : {
                        name          : item.name,
                        icon          : item.icon,
                        form          : item.form,
                        style         : item.style || {},
                        uuid          : this._vm.$genid(),
                        originalCoords: item.originalCoords,
                      }
                    }, { root: true })

                  })
              }

            })

          })

          const rowOptions = JSON.parse(JSON.stringify( rootState[ payload.from ][ curr_row ][ 'options' ] ))
          dispatch( 'copyRowOptions', { row: curr_row, rowOptions: rowOptions } )

          const rowVisibility = rootState[ payload.from ][ curr_row ][ 'active' ] ? true : false
          dispatch( 'ui/setRowActive', { row: curr_row, active: rowVisibility }, { root: true })
        })

        const gridOptions = JSON.parse(JSON.stringify( rootState[ payload.from ][ 'grid' ][ 'options' ] ))
        dispatch( 'copyOptions', { gridOptions: gridOptions })

        resolve()
      })
    },

    /**
     * Copy grid options from one module to another.
     *
     * @param {Object} rootState Access to global state.
     * @param {function} commit Synchronous invoke mutation.
     * @param {Object} payload Object that represents data sends to store
     *    - gridOptions {object} Options object to be assign.
     */
    copyOptions({ rootState, commit }, payload ) {
      return new Promise (( resolve )  => {
        const currentStoreModule = rootState.general.currentStoreModule

        commit( `${ currentStoreModule }/COPY_GRID_OPTIONS`, payload, { root: true })

        resolve()
      })
    },

    /**
     * Copy row options for given row from one module to another.
     *
     * @param {Object} rootState Access to global state.
     * @param {function} commit Synchronous invoke mutation.
     * @param {Object} payload Object that represents data sends to store
     *    - row {string} name of the row.
     *    - rowOptions {object} Options object to be assign.
     */
    copyRowOptions({ rootState, commit }, payload ) {
      return new Promise(( resolve ) => {
        const currentStoreModule = rootState.general.currentStoreModule

        commit( `${ currentStoreModule }/COPY_ROW_OPTIONS`, payload, { root: true })

        resolve()
      })
    },

  }

}