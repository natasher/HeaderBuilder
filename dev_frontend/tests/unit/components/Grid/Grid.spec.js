import { shallowMount, createLocalVue } from '@vue/test-utils'
import merge from 'lodash.merge'
import Vuex  from 'vuex'
import Grid  from '../../../../src/components/Grid/Grid.vue'

const localVue = createLocalVue()
localVue.use( Vuex )

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      ui: {
        namespaced: true,
        getters: {
          getGridStatus: jest.fn().mockReturnValue( 'custom' )
        }
      }
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    store: createStore(),
  }

  return shallowMount( Grid, merge( defaultMountingOptions, overrides ))
}

describe('Grid.vue', () => {

  describe('snapshots:', () => {

    test('should match snapshot if grid status is other then `auto`', () => {
      const wrapper = createWrapper()

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match snapshot if grid status is `auto`', () => {
      const store = createStore({
        modules: {
          ui: {
            getters: {
              getGridStatus: jest.fn().mockReturnValue( 'auto' )
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })
  })

})