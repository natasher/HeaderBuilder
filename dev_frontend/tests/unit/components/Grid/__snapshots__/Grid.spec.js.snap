// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid.vue snapshots: should match snapshot if grid status is \`auto\` 1`] = `
<div
  class="grid"
  data-status="auto"
>
  <wrapperclearfix-stub
    class="window primary"
  >
    <gridheader-stub />
     
    <wrapperclearfix-stub
      class="main"
    >
      <actionbar-stub />
       
      <firstrow-stub />
       
      <secondrow-stub />
    </wrapperclearfix-stub>
     
    <div
      class="grid-auto"
    >
      
      You're in Auto Mode
    
    </div>
  </wrapperclearfix-stub>
</div>
`;

exports[`Grid.vue snapshots: should match snapshot if grid status is other then \`auto\` 1`] = `
<div
  class="grid"
  data-status="custom"
>
  <wrapperclearfix-stub
    class="window primary"
  >
    <gridheader-stub />
     
    <wrapperclearfix-stub
      class="main"
    >
      <actionbar-stub />
       
      <firstrow-stub />
       
      <secondrow-stub />
    </wrapperclearfix-stub>
     
    <!---->
  </wrapperclearfix-stub>
</div>
`;
